### PR TITLE
feat(launchpad): stream wallet events to front-end

### PR DESCRIPTION
**Pull request by alex1307**
Description

Provide a new Tari event stream for:
	- transaction received (source: wallet)
	- transaction sent (source: wallet)
	- transaction cancelled (source: wallet)
	- transaction mined but unconfirmed (source: wallet)
	- transaction mined and confirmed (source: wallet)

Expose a streaming interface for all data we log using.
Create grpc TransactionEvent model and implement grpc streaming endpoint.
Forward logging data from notifier(app. logger) to the grpc streaming service.
Map the logging data to TransactionEvent model (Vec<String> -> TransactionEvent)

Refactoring mapping Transaction -> Vec<String> -> TransactionEvent with enum.
Refactoring EVENT_LISTENER tuple with struct Wallet Event Listener

Motivation and Context
The launchpad users want to see the actual mining progress and all mining relalted events on the UI.
Hence we should forward the data/log streams from the underlying apps to the front-end.

How Has This Been Tested?

!!! TO_DO: The change IS NOT TESTED.
The change should be tested manually along with base node, tor, and launchpad.

### CODE REVIEW

File: applications/tari_console_wallet/src/grpc/mod.rs [hunk 1]
```diff
@@ -3,4 +3,50 @@
...
+impl TryFrom<WalletEventMessage> for TransactionEvent {
+    type Error = String;
+
+    fn try_from(value: WalletEventMessage) -> Result<Self, Self::Error> {
+        match value {
+            WalletEventMessage::Completed((event, tx)) => Ok(TransactionEvent {
+                event,
+                tx_id: tx.tx_id.to_string(),
+                source_pk: tx.source_public_key.to_hex().into_bytes(),
+                dest_pk: tx.destination_public_key.clone().to_hex().into_bytes(),
```
Comment (CjS77) [inline applications/tari_console_wallet/src/grpc/mod.rs:23]:
```suggestion
                dest_pk: tx.destination_public_key.to_hex().into_bytes(),
```

File: applications/tari_console_wallet/src/grpc/mod.rs [hunk 2]
```diff
@@ -3,4 +3,50 @@
...
+                status: tx.status.to_string(),
+                direction: tx.direction.to_string(),
+                amount: tx.amount.as_u64(),
+                message: tx.message,
+            }),
+            WalletEventMessage::Outbound((event, outbound)) => Ok(TransactionEvent {
+                event,
+                tx_id: outbound.tx_id.to_string(),
+                source_pk: vec![],
+                dest_pk: outbound.destination_public_key.clone().to_hex().into_bytes(),
```
Comment (CjS77) [inline applications/tari_console_wallet/src/grpc/mod.rs:33]:
```suggestion
                dest_pk: outbound.destination_public_key.to_hex().into_bytes(),
```

File: applications/tari_console_wallet/src/grpc/mod.rs [hunk 3]
```diff
@@ -3,4 +3,50 @@
...
+                status: outbound.status.to_string(),
+                direction: "outbound".to_string(),
+                amount: outbound.amount.as_u64(),
+                message: outbound.message,
+            }),
+            WalletEventMessage::Inbound((event, inbound)) => Ok(TransactionEvent {
+                event,
+                tx_id: inbound.tx_id.to_string(),
+                source_pk: inbound.source_public_key.clone().to_hex().into_bytes(),
```
Comment (CjS77) [inline applications/tari_console_wallet/src/grpc/mod.rs:42]:
```suggestion
                source_pk: inbound.source_public_key.to_hex().into_bytes(),
```

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 4]
```diff
@@ -542,6 +551,41 @@ impl wallet_server::Wallet for WalletGrpcServer {
...
+        &self,
+        _request: tonic::Request<TransactionEventRequest>,
+    ) -> Result<Response<Self::StreamTransactionEventsStream>, Status> {
+        let (mut sender, receiver) = mpsc::channel(100);
+        task::spawn(async move {
+            let event_listener = &WALLET_EVENT_LISTENER.receiver;
+            loop {
+                let transaction_event = match event_listener.recv_timeout(Duration::from_secs(1)) {
+                    Ok(source) => TransactionEvent::try_from(source),
+                    Err(_) => continue,
```
Comment (CjS77) [inline applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs:564]: This is essentially a spin-lock of sorts. Stanimal - is this the best practice approach here?

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 5]
```diff
@@ -542,6 +551,41 @@ impl wallet_server::Wallet for WalletGrpcServer {
...
+                };
+                if transaction_event.is_ok() {
+                    let response = TransactionEventResponse {
+                        transaction: Some(transaction_event.unwrap()),
+                    };
+                    match sender.send(Ok(response)).await {
+                        Ok(_) => (),
+                        Err(err) => {
+                            warn!(target: LOG_TARGET, "Error sending transaction via GRPC:  {}", err);
+                            match sender.send(Err(Status::unknown("Error sending data"))).await {
```
Comment (CjS77) [inline applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs:574]: You've just failed to send a message. Now you're trying to send the error status to the very thing that just failed.

Is this intentional?

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 6]
```diff
@@ -542,6 +551,41 @@ impl wallet_server::Wallet for WalletGrpcServer {
...
+                                Ok(_) => (),
+                                Err(send_err) => {
+                                    warn!(target: LOG_TARGET, "Error sending error to GRPC client: {}", send_err)
+                                },
+                            }
+                            return;
+                        },
```
Comment (CjS77) [inline applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs:581]: Can be tidied up. (Clippy should catch this)

```suggestion
if let Err(err) = sender.send(Ok(response)).await {
                        warn!(target: LOG_TARGET, "Error sending transaction via GRPC:  {}", err);
                        if let Err(send_err) = sender.send(Err(Status::unknown("Error sending data"))).await {
                            warn!(target: LOG_TARGET, "Error sending error to GRPC client: {}", send_err)
                        }
                        return;
                    }
```

**Review by CjS77 — COMMENTED** [review]
Seems ok, but I'd like @stringhandler or @stanimal to review to check whether the standard Tari patterns & best practices are being employed

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs (hunk 4 above)
Comment (stringhandler) [reply applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs:564]: You should be able to remove the timeout and use recv

File: applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs (hunk 5 above)
Comment (stringhandler) [reply applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs:574]: We do this in other places in other GRPC with streams. At worst case, you fail twice, at best, you get the error to the client. I think in most cases it ends up being a double fail

**Review by tarnas14 — CHANGES_REQUESTED** [review]
~@alex1307  there are unverified commits in this PR, I don't think we can allow this merge unless you change it~
nvm they're signed but emails don't match, so there's no problem, I think

**Comment by alex1307** [comment]
The PR has unsigned commits. New PR is created.

**Pull request closed without merging**